### ISSUE TITLE
Investigate transient invalid state in map and array vectors

### ISSUE THREAD

**Issue by bikramSingh91** [facebookincubator/velox#3226]
A recent change (#3078) attempted to add validity checks for map and array vectors which checked that the values inside offset and size buffers pointed to valid locations in the elements/keys/values vectors.

This resulted in a lot (~180) of test failures where this check was violated.
Therefore, filing this issue to investigate the nature of these violations and fix them.

**Comment by bikramSingh91** [comment]
cc: @Yuhta @oerling

**Comment by bikramSingh91** [comment]
I ran all the tests on my local and tried to narrow down the root cause for each.

TLDR: the issue is that the API for array and maps allows mutable access to the inner keys/values and this culminated to a pattern where the vector is created first in an invalid state and later the right values and sizes are populated into its child vectors. Some of the failures have easy fixes and some don't, but even after we fix all of them, there is no guarantee that the vector will never get into an invalid state since those underlying vectors or buffers can still be manipulated after construction.

Suggestion: This might a necessary evil since direct access to underlying state is necessary for manipulation used across the code base. What we can perhaps do it use this consistency check (mentioned in the top post) as a utility method that we can use as a DCHECK() in places where we want to be sure they make sense before we operate on a vector.

Details on some of the failed tests:

- velox_vector_test
Reason:
Looks like it uses BaseVector::create() to create a fixed size array that has pre-filled offsets and sizes but an empty values vector. This values vector is then filled later on by copying values from another array.

`[  FAILED  ]  VectorTest, fixedSizeArray`


- velox_dwio_dwrf_column_reader_test
Reason:
creating an array then manipulating the values later. In MapColumnReader::next() and ListColumnReader::next() at the end it checks for Children the recursively fills it
    
```
    115: [  FAILED  ] 24 tests, listed below:
    115: [  FAILED  ] TestColumnReaderNoReuse/TestColumnReader.testList/0, where GetParam() = 2-byte object <00-00> => creating an array then manipulating the values later
    115: [  FAILED  ] TestColumnReaderNoReuse/TestColumnReader.testMap/0, where GetParam() = 2-byte object <00-00>
    115: [  FAILED  ] TestColumnReaderNoReuse/TestColumnReader.testLargeSkip/0, where GetParam() = 2-byte object <00-00>
    115: [  FAILED  ] TestColumnReaderNoReuse/TestColumnReader.testListVectorTypeChange/0, where GetParam() = 2-byte object <00-00>
    115: [  FAILED  ] TestColumnReaderNoReuse/TestColumnReader.testMapVectorTypeChange/0, where GetParam() = 2-byte object <00-00>
    115: [  FAILED  ] TestColumnReaderReuse/TestColumnReader.testList/0, where GetParam() = 2-byte object <00-01>
    115: [  FAILED  ] TestColumnReaderReuse/TestColumnReader.testMap/0, where GetParam() = 2-byte object <00-01>
    115: [  FAILED  ] TestColumnReaderReuse/TestColumnReader.testLargeSkip/0, where GetParam() = 2-byte object <00-01>
    115: [  FAILED  ] TestColumnReaderReuse/TestColumnReader.testListVectorTypeChange/0, where GetParam() = 2-byte object <00-01>
    115: [  FAILED  ] TestColumnReaderReuse/TestColumnReader.testMapVectorTypeChange/0, where GetParam() = 2-byte object <00-01>
    115: [  FAILED  ] TestSelectiveColumnReader/TestColumnReader.testListVectorTypeChange/0, where GetParam() = 2-byte object <01-00>
    115: [  FAILED  ] TestSelectiveColumnReader/TestColumnReader.testMapVectorTypeChange/0, where GetParam() = 2-byte object <01-00>
    115: [  FAILED  ] TestNonSelectiveColumnReaderNoReuse/TestNonSelectiveColumnReader.testListWithNulls/0, where GetParam() = false
    115: [  FAILED  ] TestNonSelectiveColumnReaderNoReuse/TestNonSelectiveColumnReader.testListSkipWithNulls/0, where GetParam() = false
    115: [  FAILED  ] TestNonSelectiveColumnReaderNoReuse/TestNonSelectiveColumnReader.testListSkipWithNullsNoData/0, where GetParam() = false
    115: [  FAILED  ] TestNonSelectiveColumnReaderNoReuse/TestNonSelectiveColumnReader.testMapWithNulls/0, where GetParam() = false
    115: [  FAILED  ] TestNonSelectiveColumnReaderNoReuse/TestNonSelectiveColumnReader.testMapSkipWithNulls/0, where GetParam() = false
    115: [  FAILED  ] TestNonSelectiveColumnReaderNoReuse/TestNonSelectiveColumnReader.testMapSkipWithNullsNoData/0, where GetParam() = false
    115: [  FAILED  ] TestNonSelectiveColumnReaderReuse/TestNonSelectiveColumnReader.testListWithNulls/0, where GetParam() = true
    115: [  FAILED  ] TestNonSelectiveColumnReaderReuse/TestNonSelectiveColumnReader.testListSkipWithNulls/0, where GetParam() = true
    115: [  FAILED  ] TestNonSelectiveColumnReaderReuse/TestNonSelectiveColumnReader.testListSkipWithNullsNoData/0, where GetParam() = true
    115: [  FAILED  ] TestNonSelectiveColumnReaderReuse/TestNonSelectiveColumnReader.testMapWithNulls/0, where GetParam() = true
    115: [  FAILED  ] TestNonSelectiveColumnReaderReuse/TestNonSelectiveColumnReader.testMapSkipWithNulls/0, where GetParam() = true
    115: [  FAILED  ] TestNonSelectiveColumnReaderReuse/TestNonSelectiveColumnReader.testMapSkipWithNullsNoData/0, where GetParam() = true
```

    
- Velox_dwrf_column_writer_test
Reason is same as before:
creating an array then manipulating the values later. In MapColumnReader::next() and ListColumnReader::next() at the end it checks for Children the recursively fills it
    
- velox_dwrf_e2e_writer_test
Reason is same as before:
creating an array then manipulating the values later. In MapColumnReader::next() and ListColumnReader::next() at the end it checks for Children the recursively fills it

```
    122: [  FAILED  ] E2EWriterTests.E2E
    122: [  FAILED  ] E2EWriterTests.FlatMapDictionaryEncoding
    122: [  FAILED  ] E2EWriterTests.fuzzComplex
    122: [  FAILED  ] E2EWriterTests.fuzzFlatmap
    122: [  FAILED  ] E2EEncryptionTest.EncryptSelectedFields
```
    
- velox_aggregates_test
Reason:
Wrong size of values vector create, the offset is 0 and size 3 but the values vector is a dict of size 1

```
        132: [  FAILED  ] ApproxPercentileTest.invalidEncoding
```
        
    
- Velox_exec_Test
Reason:
PrestoSerializer::readMapVector creates a map vector then manipulates the offsets and sizes later on.

```
    139: [  PASSED  ] 471 tests.
    139: [  FAILED  ] 22 tests, listed below:
    139: [  FAILED  ] SpillerTest/NoHashJoin.spilFew/0, where GetParam() = 8-byte object <00-00 00-00 00-00 00-00>
    139: [  FAILED  ] SpillerTest/NoHashJoin.spilFew/1, where GetParam() = 8-byte object <00-00 00-00 08-00 00-00>
    139: [  FAILED  ] SpillerTest/NoHashJoin.spilFew/2, where GetParam() = 8-byte object <03-00 00-00 00-00 00-00>
    139: [  FAILED  ] SpillerTest/NoHashJoin.spilFew/3, where GetParam() = 8-byte object <03-00 00-00 08-00 00-00>
    139: [  FAILED  ] SpillerTest/NoHashJoin.spilMost/0, where GetParam() = 8-byte object <00-00 00-00 00-00 00-00>
    139: [  FAILED  ] SpillerTest/NoHashJoin.spilMost/1, where GetParam() = 8-byte object <00-00 00-00 08-00 00-00>
    139: [  FAILED  ] SpillerTest/NoHashJoin.spilMost/2, where GetParam() = 8-byte object <03-00 00-00 00-00 00-00>
    139: [  FAILED  ] SpillerTest/NoHashJoin.spilMost/3, where GetParam() = 8-byte object <03-00 00-00 08-00 00-00>
    139: [  FAILED  ] SpillerTest/NoHashJoin.spillAll/0, where GetParam() = 8-byte object <00-00 00-00 00-00 00-00>
    139: [  FAILED  ] SpillerTest/NoHashJoin.spillAll/1, where GetParam() = 8-byte object <00-00 00-00 08-00 00-00>
    139: [  FAILED  ] SpillerTest/NoHashJoin.spillAll/2, where GetParam() = 8-byte object <03-00 00-00 00-00 00-00>
    139: [  FAILED  ] SpillerTest/NoHashJoin.spillAll/3, where GetParam() = 8-byte object <03-00 00-00 08-00 00-00>
    139: [  FAILED  ] SpillerTest/NoHashJoin.error/0, where GetParam() = 8-byte object <00-00 00-00 00-00 00-00>
    139: [  FAILED  ] SpillerTest/NoHashJoin.error/1, where GetParam() = 8-byte object <00-00 00-00 08-00 00-00>
    139: [  FAILED  ] SpillerTest/NoHashJoin.error/2, where GetParam() = 8-byte object <03-00 00-00 00-00 00-00>
    139: [  FAILED  ] SpillerTest/NoHashJoin.error/3, where GetParam() = 8-byte object <03-00 00-00 08-00 00-00>
    139: [  FAILED  ] SpillerTest/AllTypes.nonSortedSpillFunctions/0, where GetParam() = 8-byte object <00-00 00-00 00-00 00-00>
    139: [  FAILED  ] SpillerTest/AllTypes.nonSortedSpillFunctions/1, where GetParam() = 8-byte object <00-00 00-00 08-00 00-00>
    139: [  FAILED  ] SpillerTest/AllTypes.nonSortedSpillFunctions/2, where GetParam() = 8-byte object <01-00 00-00 00-00 00-00>
    139: [  FAILED  ] SpillerTest/AllTypes.nonSortedSpillFunctions/3, where GetParam() = 8-byte object <01-00 00-00 08-00 00-00>
    139: [  FAILED  ] SpillerTest/AllTypes.nonSortedSpillFunctions/6, where GetParam() = 8-byte object <03-00 00-00 00-00 00-00>
    139: [  FAILED  ] SpillerTest/AllTypes.nonSortedSpillFunctions/7, where GetParam() = 8-byte object <03-00 00-00 08-00 00-00>
```

**Comment by Yuhta** [comment]
`ApproxPercentileTest.invalidEncoding` this is a typo, I can fix it
